design/object-bucket/object-bucket.md

```diff
@@ -43,115 +51,159 @@ Add a CephObjectBucketClaim API endpoint and control loop to the Rook-Ceph opera
 
 ![admin-actions](./cobc-admin.png)
 
-1. The admin creates a StorageClass with `parameters[serviceName]=OBJECT-STORE-SERVICE` and `parameters[serviceNamespace]=OBJECT-STORE-NAMESPACE`
+1. The admin creates a StorageClass with the `provisioner=ceph.rook.io/object`, `parameters[serviceName]=OBJECT-STORE-SERVICE`, and `parameters[serviceNamespace]=OBJECT-STORE-NAMESPACE` key-values.
 
 **Use Case: Provision a Bucket** 
 
-_As a Kubernetes user, I want to leverage the Kubernetes API to create Ceph Object Buckets. I expect to get back the bucket connection information in a Pod-attachable object._
+_As a Kubernetes user, I want to leverage the Kubernetes API to create object store buckets. I expect to get back the bucket endpoint information in a Pod-consumable API resource._
+ 
+//TODO update diagram to reflect API changes  
  
 ![user-actions](./cobc-user.png)
 
-1. The user creates a CephObjectBucketClaim in their namespace.
-1. The operator detects the new ObjectBucketClaim instance.
-    1. The operator derives the object store service via `storageClass.parameters.objectStoreNamespace` + `storageClass.parameters.objectStore`.
-1. If a CephObjectUser for the CephObjectStore does not exist, the operator creates it and waits for the associated Secret.  This will give the operator **bucket** CREATE/GET/DELETE permissions in the object store.  This key is reused later and exists for the lifetime of the object store.
-1. The operator creates the bucket using the service endpoint and CephObjectStoreUser **bucket** access key.  The bucket name will be a concatenation of the  claim's `metadata.namespace` and `metadata.name`.  This protects against bucket naming collisions.
-    1. If the bucket already exists in the store, the operator will instead regenerate the ConfigMap and Secret associated with that bucket.  This allows users the ability to a) reacquire buckets who's claims have been deleted and b) incorporate brownfield buckets created outside of the operator.
-1. The operator creates or updates a ConfigMap in the namespace of the CephObjectBucketClaim with relevant connection data for that bucket.
-    1. The operator uses the radosgw-admin CLI to generate a **user** access key.  This key will be bound to the bucket with an **object** PUT/GET/DELETE ACL.
-    1. The operator binds the **user** key to the bucket with an **object** PUT/GET/DELETE ACL.
-1. The operator creates a Secret in the namespace of the CephObjectBucketClaim with the **user** key pair.
+1. The user creates an OBC in their namespace.
+1. The operator detects the new OBC instance.
+    1. The operator checks the `provisioner` field of the OBC's StorageClass.  If it matches, continue to the next step. Else, end the process here.
+    1. The operator gets the object store service using `storageClass.parameters.objectStoreNamespace` + `storageClass.parameters.objectStore`.
+1. The operator creates a CephObjectUser if it does not exist in order to generate the operator's access key.
```

Comment:
Do we want to still call the user a `CephObjectUser`, or are we planning on a generic object-store user claim in a later phase?

Also, maybe mention that if >1 object-store user exists in the object-store's ns then the 1st one is used.

Reply:
These are 2 different objects representing 2 different layers of abstraction.  *Claim represents the generalized API that's TBD.  CephObjectStoreUser is ceph specific.  

> Also, maybe mention that if >1 object-store user exists in the object-store's ns then the 1st one is used.

I think instead we should key off an annotation like:
```
annotations:
  createByController: ceph.rook.io/object
```
I don't think we want the operator using an admin created CephObjectStoreUser.

In the case of >1 operator created COUs, I can only see that happening if 2 ceph operators are running and hit a race condition where both create a COU within a small window.  We may want to handle this with Expectations.